betatest/forms.py

```diff
@@ -0,0 +1,199 @@
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Forms for the Instance Manager beta test
+"""
+
+# Imports #####################################################################
+
+from django import forms
+from django.contrib.auth.models import User
+from django.db import transaction
+from django.utils.text import capfirst
+from djng.forms import NgFormValidationMixin, NgModelForm, NgModelFormMixin
+from zxcvbn import password_strength
+
+from betatest.models import BetaTestApplication
+from userprofile.models import UserProfile
+
+
+# Forms #######################################################################
+
+class BetaTestApplicationForm(NgModelFormMixin, NgFormValidationMixin, NgModelForm):
+    """
+    Application form for beta testers. Creates instances of User, UserProfile,
+    and BetaTestApplication models on submit.
+    """
+    class Meta: #pylint: disable=missing-docstring
+        model = BetaTestApplication
+        exclude = ('user', 'status')
+        widgets = {
+            'public_contact_email': forms.widgets.EmailInput(attrs={
+                'validate-email': True,
+            }),
+        }
+
+    full_name = forms.CharField(
+        max_length=255,
+        help_text='Example: Albus Dumbledore',
+    )
+    username = forms.RegexField(
+        regex=r'^[\w.+-]+$',
+        max_length=30,
+        help_text=('This would also be the username of the administrator '
+                   'account on the Open edX instance.'),
+        error_messages={
+            'invalid': ('Usernames may contain only letters, numbers, and '
+                        './+/-/_ characters.'),
+            'unique': 'This username is already taken.',
+        },
+    )
+    email = forms.EmailField(
+        help_text=('This is also your account name, and where we will send '
+                   'important notices.'),
+        widget=forms.widgets.EmailInput(attrs={'validate-email': True}),
+    )
+    password = forms.CharField(
+        widget=forms.PasswordInput,
+        help_text=('Pick a password for your OpenCraft account. You will be '
+                   'able to use it to login and access your account.'),
+    )
+    password_confirmation = forms.CharField(
+        widget=forms.PasswordInput,
+        help_text=('Please use a strong password: avoid common patterns and '
+                   'make it long enough to be difficult to crack.'),
```

Comment:
I strongly prefer entropy-based password validation, using a tool like `zxcvbn`. What most websites do that use it is simply show a client-side "strength bar" while the user enters their password. Then, as they type every single character in their password, they can instantly see if the password is strong enough or not, and they can get an intuitive sense of what makes a strong password.

In such cases, it is typically not necessary to provide any further guidance to the user. People are used to making up passwords and generally have an idea of how to make a password stronger. I don't think it hurts to provide some specific examples though, such as something like "An example of a strong password would be thinking of a sentence with eight or more words and using the first letter of each word in the sentence, e.g. 'I love taking online courses with Open edX!' becomes 'IltocwOe!'"